M3-5588 Drawer component story (#8063)
linode/manager: packages/manager/src/components/Drawer/Drawer.stories.mdx

import Button from 'src/components/Button';
import Typography from 'src/components/core/Typography';
import ActionsPanel from '../ActionsPanel';
import TextField from '../TextField';
import Drawer from './Drawer';
import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs';

export const decorators = [
  (Story) => {
    const [open, setOpen] = React.useState(false);
    return Story({ open, setOpen });
  },
];

<Meta title="Components/Drawer" component={Drawer} />

# Drawer

#### Overview

- Drawers are essentially modal dialogs that appear on the right of the screen rather than the center.
- Like traditional modals, they block interaction with the page content.
- They are elevated above the app’s UI and don’t affect the screen’s layout grid.

#### Behavior

- Clicking a button on the screen opens the drawer.

> - Clicking off of the drawer closes it and data will not be saved.

Author: True! Removed.

        Click to open Drawer
      </Button>
      <Drawer title="My Drawer" {...args} open={context.open} onClose={close}>
        <Typography>
          This is some test copy which acts as content for this Drawer
          component. It's very interesting and you should read all of it. This
          text has to be sufficiently long to test that it doesn't expand the
          drawer to an unreasonable width.
        </Typography>
        <TextField
          label="Input Some Text"
          placeholder="This is a placeholder"
        />
        <ActionsPanel>
          <Button buttonType="primary" onClick={close}>
            Save
          </Button>
          <Button buttonType="secondary" onClick={close}>
            Cancel
          </Button>
        </ActionsPanel>
      </Drawer>
    </>
  );
};

<Canvas>
  <Story name="Default" decorators={decorators}>
    {Template.bind({})}
  </Story>
</Canvas>

### Component API

<hr />

_Props extends MUI's DrawerProps_

<Canvas>
  <Story
    name="Component API"
    decorators={decorators}
    args={{
      open: false,
      onClose: () => null,
      title: 'My Drawer',
      wide: false,
    }}
    argTypes={{
      open: {

>         description: 'Whether or not the Drawer is currently open',

Author: I tried a few things to get around this, but wasn't able to find a good way to address it.

import Button from 'src/components/Button';
import Typography from 'src/components/core/Typography';
import ActionsPanel from '../ActionsPanel';
import TextField from '../TextField';
import Drawer from './Drawer';
import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs';

export const decorators = [
  (Story) => {
    const [open, setOpen] = React.useState(false);
    return Story({ open, setOpen });
  },
];

<Meta title="Components/Drawer" component={Drawer} />

# Drawer

#### Overview

- Drawers are essentially modal dialogs that appear on the right of the screen rather than the center.
- Like traditional modals, they block interaction with the page content.
- They are elevated above the app’s UI and don’t affect the screen’s layout grid.

#### Behavior

- Clicking a button on the screen opens the drawer.

> - Drawers can also be closed by clicking the “X” icon or the “Close” button.

Reviewer: Include `esc` key for closing drawers

```suggestion
- Drawers can be closed by pressing the `esc` key, clicking the “X” icon, or clicking the “Cancel” button.
```

export const Template = (args, context) => {
  const close = () => context.setOpen(false);
  return (
    <>
      <Button buttonType="primary" onClick={() => context.setOpen(true)}>
        Click to open Drawer
      </Button>
      <Drawer title="My Drawer" {...args} open={context.open} onClose={close}>
        <Typography>
          This is some test copy which acts as content for this Drawer
          component. It's very interesting and you should read all of it. This
          text has to be sufficiently long to test that it doesn't expand the
          drawer to an unreasonable width.
        </Typography>
        <TextField
          label="Input Some Text"
          placeholder="This is a placeholder"
        />
        <ActionsPanel>

>           <Button buttonType="primary" onClick={close}>
>             Save
>           </Button>
>           <Button buttonType="secondary" onClick={close}>
>             Cancel
>           </Button>

Reviewer: Update button order so action button is on the right

```suggestion
          <Button buttonType="secondary" onClick={close}>
            Cancel
          </Button>
          <Button buttonType="primary" onClick={close}>
            Save
          </Button>
```

        }
      },
      onClose: {
        description: 'The handler called when the "X" button is clicked',
        type: {
          required: true
        }
      },
      title: {
        description: 'The title displayed at the top of the Drawer',
        type: {
          required: true
        }
      },
      wide: {
        description:
          'Whether or not the drawer should be wide (at small viewports and above)',
        table: {
          defaultValue: { summary: 'false' },
        },
      },
    }}
    parameters={{ docs: { source: { code: DrawerSourceCode } } }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable
  story="Component API"
  include={['open', 'onClose', 'title', 'wide']}
/>

export const DrawerSourceCode = `
  <Drawer title="My Drawer" open={open} onClose={close}>
  <Typography>
    This is some test copy which acts as content for this Drawer
    component. It's very interesting and you should read all of it. This
    text has to be sufficiently long to test that it doesn't expand the
    drawer to an unreasonable width.
  </Typography>
  <TextField
    label="Input Some Text"
    placeholder="This is a placeholder"
  />
  <ActionsPanel>

>     <Button buttonType="primary" onClick={close}>
>       Save
>     </Button>
>     <Button buttonType="secondary" onClick={close}>
>       Cancel
>     </Button>

Reviewer: Update button order so action button is on the right here also

```suggestion
    <Button buttonType="secondary" onClick={close}>
      Cancel
    </Button>
    <Button buttonType="primary" onClick={close}>
      Save
    </Button>
```

export const Template = (args, context) => {
  const close = () => context.setOpen(false);
  return (
    <>
      <Button buttonType="primary" onClick={() => context.setOpen(true)}>
        Click to open Drawer
      </Button>
      <Drawer title="My Drawer" {...args} open={context.open} onClose={close}>
        <Typography>
          This is some test copy which acts as content for this Drawer
          component. It's very interesting and you should read all of it. This
          text has to be sufficiently long to test that it doesn't expand the
          drawer to an unreasonable width.
        </Typography>
        <TextField
          label="Input Some Text"
          placeholder="This is a placeholder"
        />
        <ActionsPanel>
          <Button buttonType="secondary" onClick={close}>
            Cancel
          </Button>
          <Button buttonType="primary" onClick={close}>
            Save
          </Button>
        </ActionsPanel>
      </Drawer>
    </>
  );
};

<Canvas>
  <Story
    name="Default"
    decorators={decorators}
    parameters={{ docs: { source: { code: DrawerSourceCode } } }}
  >
    {Template.bind({})}
  </Story>

> </Canvas>

Reviewer 2: What about removing this and renaming the `Component API` story to `Default` since they basically seem to be the same? This would also cleanup the sidebar a bit (specifically in Canvas view, there is no difference between `Default` and `Component API`)

Reviewer 3: It seems like the code generated from the ArgsTable never updates regardless of the changes made under the `Component API` example so I'm okay with moving it up. In that case, would we move the other examples below the ArgsTable (ie. for Buttons: primary, ArgsTable, secondary, outlined)?

Author: You're right that the code in the ArgsTable doesn't change, but the component does change (e.g. if you toggle "Wide" in the ArgsTable, the drawer will be updated accordingly), which is why I think a separate component specifically for the ArgsTable makes sense. It's less clear in this example with the Drawer since you have to click a button in order to see the drawer.

(Also, I'd love to eventually find a way to update the code when values in the ArgsTable changes!)
